File: 876.链表的中间结点.go
/*
 * @lc app=leetcode.cn id=876 lang=golang
 *
 * [876] 链表的中间结点
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func middleNode(head *ListNode) *ListNode {
	// dummyNode := head
	// nodeCount := 0
	// for head != nil {
	//     nodeCount++
	//     head = head.Next
	// }
	// nodeCount /= 2
	// for nodeCount != 0 {
	//     dummyNode = dummyNode.Next
	//     nodeCount--
	// }
	// return dummyNode

	// 快慢指针
	slow := head
	for head != nil && head.Next != nil {
		slow = slow.Next
		head = head.Next.Next
	}
	return slow
}

// @lc code=end

File: 876.链表的中间结点.py
#
# @lc app=leetcode.cn id=876 lang=python3
#
# [876] 链表的中间结点
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def middleNode(self, head: ListNode) -> ListNode:
        # dummy_node = head
        # node_count = 0
        # while head:
        #     node_count += 1
        #     head = head.next
        # node_count = node_count // 2
        # while node_count != 0:
        #     dummy_node = dummy_node.next
        #     node_count -= 1
        # return dummy_node
        # 快慢指针
        pre_node = head
        while head and head.next:
            pre_node = pre_node.next
            head = head.next.next
        return pre_node
# @lc code=end

